// bobflux/dist/index.js
define(["require", "exports", 'node_modules/bobril/index', 'node_modules/fun-model/dist/index', 'node_modules/fun-model/dist/index', './component'], function (require, exports, b, index_1, index_2, component_1) {
    function __export(m) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }
    exports.getState = index_2.getState;
    exports.setState = index_2.setState;
    exports.rootCursor = index_2.rootCursor;
    exports.shallowCopy = index_2.shallowCopy;
    exports.createAction = index_2.createAction;
    exports.createActions = index_2.createActions;
    __export(component_1);
    exports.bootstrap = function (defaultState, debugCallback, subStateSeparator) {
        if (debugCallback === void 0) { debugCallback = undefined; }
        if (subStateSeparator === void 0) { subStateSeparator = '.'; }
        index_1.bootstrap(defaultState, function () { return b.invalidate(); }, function (message, params) { debugCallback && debugCallback("bobflux -> " + message, params); });
    };
});
